pkg/metrics/collector/sockstat.go

```diff
@@ -0,0 +1,185 @@
+			"Number of sockets in time wait state",
+			nil, nil,
+		),
+		socketMemoryDesc: prometheus.NewDesc(
+			"socket_memory",
+			"Amount of memory used by sockets",
+			nil, nil,
+		),
+	}, nil
+}
+
+type socketStats struct {
+	tcpInUse        uint64 // num of tcp in use sockets
+	udpInUse        uint64 // num of udp in use sockets
+	tcpTimeWait     uint64 // num of tcp timewait sockets
+	memUsedInBlocks uint64 // sum of tcp and udp memory used in blocks
+}
+
+func getSockStats(fileName string) (*socketStats, error) {
+	file, err := os.Open(fileName)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	return parseSockStats(file)
+}
+
+func (s *socketStats) merge(o *socketStats) {
+	s.tcpInUse += o.tcpInUse
+	s.udpInUse += o.udpInUse
+	s.tcpTimeWait += o.tcpTimeWait
+	s.memUsedInBlocks += o.memUsedInBlocks
+}
+
+func extractSockStatUint(field string, s string, out interface{}) error {
+	v, err := strconv.ParseUint(s, 10, 64)
+	if err != nil {
+		return fmt.Errorf("invalid value %q for field %q in sockstats", s, field)
+	}
+	outValue := out.(*uint64)
+	*outValue = v
+	return nil
+}
+
+type sockStatEntry struct {
+	out     interface{}
+	extract func(string, string, interface{}) error
+}
+
+func extractSockStatFields(fields []string, entries map[string]*sockStatEntry) error {
+	for i := 0; i < len(fields) && i+1 < len(fields); i += 2 {
+		k, v := fields[i], fields[i+1]
+		if e, ok := entries[k]; !ok {
+			continue
+		} else if err := e.extract(k, v, e.out); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func parseLine(in string) (*socketStats, error) {
+	line := strings.Split(in, " ")
+	if len(line) < 2 || len(line[0]) < 2 || line[0][len(line[0])-1] != ':' {
+		return nil, fmt.Errorf("invalid line %q", in)
+	}
+
+	ret := &socketStats{}
+	tcpEntries := map[string]*sockStatEntry{
+		"inuse": {&ret.tcpInUse, extractSockStatUint},
+		"tw":    {&ret.tcpTimeWait, extractSockStatUint},
+		"mem":   {&ret.memUsedInBlocks, extractSockStatUint},
+	}
+
+	udpEntries := map[string]*sockStatEntry{
+		"inuse": {&ret.udpInUse, extractSockStatUint},
+		"mem":   {&ret.memUsedInBlocks, extractSockStatUint},
+	}
+
+	protocol := line[0][:len(line[0])-1]
+	rest := line[1:]
+
+	switch protocol {
+	case "TCP", "TCP6":
+		if err := extractSockStatFields(rest, tcpEntries); err != nil {
+			return nil, err
+		}
+	case "UDP", "UDP6":
+		if err := extractSockStatFields(rest, udpEntries); err != nil {
+			return nil, err
+		}
+	}
+	return ret, nil
+}
+
+func parseSockStats(r io.Reader) (*socketStats, error) {
+	var lineCount int = 0
+
+	scanner := bufio.NewScanner(r)
+	ret := &socketStats{}
+	for scanner.Scan() {
+		lineCount++
+		if s, err := parseLine(scanner.Text()); err == nil {
+			ret.merge(s)
+		} else {
+			return nil, err
+		}
+	}
+	if lineCount == 0 {
+		return nil, fmt.Errorf("empty file")
+	}
+	return ret, nil
+}
+
+func (c *sockStatCollector) Update(ch chan<- prometheus.Metric) error {
+	ss, err := getSockStats(procFilePath("net/sockstat"))
```

Comment:
Why not pass location of sockstat and sockstat6 as flags with default value `/proc/net/{sockstat,sockstat6}?`
What is the purpose of procFilePath?

Reply:
Because there is no point to have individual flags when many collectors share the same proc directory. The reason to have the procFilePath() is that it is easy to change proc path for all. This is used when the program is running in a container, so that the host proc can be mapped to a different directory inside the container for the program to run correctly.